src/logs.ts

```diff
@@ -0,0 +1,225 @@
+import { LOG_LEVEL } from "./constants";
+import { PrettyLogs } from "./pretty-logs";
+import { LogParams, LogReturn, Metadata, LogLevel } from "./types/log-types";
+
+export class Logs {
+  private _maxLevel = -1;
+  static console: PrettyLogs;
+
+  private _log({ level, consoleLog, logMessage, metadata, type }: LogParams): LogReturn | null {
+    if (this._getNumericLevel(level) > this._maxLevel) return null; // filter out more verbose logs according to maxLevel set in config
+
+    // needs to generate three versions of the information.
+    // they must all first serialize the error object if it exists
+    // - the comment to post on supabase (must be raw)
+    // - the comment to post on github (must include diff syntax)
+    // - the comment to post on the console (must be colorized)
+
+    consoleLog(logMessage, metadata || undefined);
+    return new LogReturn(
+      {
+        raw: logMessage,
+        diff: this._diffColorCommentMessage(type, logMessage),
+        type,
+        level,
+      },
+      metadata
+    );
+  }
+
+  private _addDiagnosticInformation(metadata: Record<string, unknown> | string | number | null | undefined) {
+    // this is a utility function to get the name of the function that called the log
+    // I have mixed feelings on this because it manipulates metadata later possibly without the developer understanding why and where,
+    // but seems useful for the metadata parser to understand where the comment originated from
+
+    if (!metadata) {
+      metadata = {};
+    }
+    if (typeof metadata == "string" || typeof metadata == "number") {
+      // TODO: think i need to support every data type
```

Comment:
Can this TODO be solved beforehand?

Reply:
Well it's not required to handle any other data types really. `Metadata` has `[key: string]: unknown;` so it can already process every data type so long as it's a property of metadata which is cleaner than allowing and having to process others isn't it?

So I think just a clean up here is needed because `_addDiagnosticInformation()` should only really be passed a `Metadata` object or undefined as that's what the public log functions accept as input.

This way invoking logs should just be `.fatal("xyz error", { error: e, a: [], b: {}, ...}`